File: Task2/apz-pzpi-21-11-milennyi-ivan-task2/src/db/services/storekeeper_service.rs
use std::sync::Arc;
use async_trait::async_trait;
use sqlx::{MySql, Pool};
use crate::db::service_error::ServiceError;
use crate::db::traits::{AuthService, Service, StorekeeperManage};
use crate::models::Storekeeper;
use crate::view_models::StorekeeperVM;

pub(crate) struct StorekeeperService<T>{
    pool: Arc<T>,
}
#[async_trait]
impl Service<Pool<MySql>> for StorekeeperService<Pool<MySql>> {
    type Model = Storekeeper;
    type Error = ServiceError;
    type ViewModel = StorekeeperVM;

    fn new(pool: Arc<Pool<MySql>>) -> Self {
        StorekeeperService { pool }
    }

    async fn create(&self, item: Self::Model) -> Result<Self::Model, Self::Error> {
        todo!()
    }

    async fn delete(&self, item_id: u64) -> Result<(), Self::Error> {
        todo!()
    }

    async fn update(&self, item: Self::Model) -> Result<Self::Model, Self::Error> {
        todo!()
    }

    async fn get_all(&self) -> Result<Vec<Self::Model>, Self::Error> {
        todo!()
    }

    async fn get_by_id(&self, id: u64) -> Result<Option<Self::Model>, Self::Error> {
        todo!()
    }
}
#[async_trait]
impl AuthService<Pool<MySql>> for StorekeeperService<Pool<MySql>> {
    async fn check_username(&self, user: &Self::Model) -> Result<bool, Self::Error> {
        todo!()
    }

    async fn authorize(&self, username: String, password_hash: String) -> Result<Option<Self::ViewModel>, Self::Error> {
        todo!()
    }
}

#[async_trait]
impl StorekeeperManage<Pool<MySql>> for StorekeeperService<Pool<MySql>>{

}

File: Task4/apz-pzpi-21-11-milennyi-ivan-task4/src/api/services/StorekeeperService.ts
import axios from "axios";
import { Storekeeper } from "../../models/Storekeeper";
import { API_URL } from "../../utils/config";
import { AuthService } from "../interfaces/IAuthService";
import { IStorekeeperService } from "../interfaces/IStorekeeperService";

export class StorekeeperService implements IStorekeeperService, AuthService<Storekeeper>{
    private static STOREKEEPER_URLS = {
        GET_ALL: `${API_URL}/storekeeper`,
        GET_BY_ID: (id: number) => `${API_URL}/storekeeper/${id}`,
        CREATE: `${API_URL}/storekeeper/create`,
        DELETE: (id: number) => `${API_URL}/storekeeper/delete/${id}`,
        UPDATE: `${API_URL}/storekeeper/update`,
        AUTHORIZE: `${API_URL}/storekeeper/authorize`
    }
    async create(item: Storekeeper): Promise<Storekeeper> {
        const response = await axios.post<Storekeeper>(StorekeeperService.STOREKEEPER_URLS.CREATE, item);
        return response.data;
    }
    async delete(itemId: number): Promise<void> {
        await axios.delete(StorekeeperService.STOREKEEPER_URLS.DELETE(itemId));
    }
    async update(item: Storekeeper): Promise<Storekeeper> {
        const response = await axios.put<Storekeeper>(StorekeeperService.STOREKEEPER_URLS.UPDATE, item);
        return response.data;
    }
    async getAll(): Promise<Storekeeper[]> {
        const response = await axios.get<Storekeeper[]>(StorekeeperService.STOREKEEPER_URLS.GET_ALL);
        return response.data;
    }
    async getById(id: number): Promise<Storekeeper | null> {
        const response = await axios.get<Storekeeper | null>(StorekeeperService.STOREKEEPER_URLS.GET_BY_ID(id));
        return response.data;
    }
    async authorize(username: string, passwordHash: string): Promise<Storekeeper | null> {
        const response = await axios.post<Storekeeper | null>(StorekeeperService.STOREKEEPER_URLS.AUTHORIZE, {username: username, password_hash: passwordHash});
        return response.data;
    }
}